torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -920,9 +895,9 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::broadcast(
   const auto& device = inputs[0].device();
   switch (device.type()) {
     case at::kCPU:
-#ifdef USE_CUDA
+      break;
     case at::kCUDA:
-#endif
+      TORCH_INTERNAL_ASSERT(c10::impl::hasDeviceGuardImpl(at::kCUDA));
```

Comment:
1) nit: use .hasCUDA()?
2) let's preserve the old behavior with invalidArgument instead of checkfail?

Reply:
> let's preserve the old behavior with invalidArgument instead of checkfail?

The reason I put an assert here is because this should never ever possibly occur: if the user has been able to give us a CUDA tensor, then CUDA _must_ be loaded and that check _must_ pass. In fact, I wouldn't even oppose removing that assert entirely, since in a sense it's pointless...

I'll replace this with `at::hasCUDA()`, which is more concise and I think will give the same answer, although technically it does behave somewhat differently: PyTorch apparently has two "registries" for CUDA to plug in, one is the CUDAHooks (which is what is checked by `at::hasCUDA()`) the other one is the DeviceGuardImpl registry (which is used by c10::Events and such, which is what we actually care about here).